Implement fully declarative approach via direct `config` as an alternative to specifying `flake`

oddlama opened the pull request:
This PR adds an alternative option `vms.<name>.config` that can be used instead of specifying `vms.<name>.flake`. This way of doing configuration doesn't impose any restrictions on the user's flake, but can only be used declaratively. This is similar to how e.g. nixos-container is doing this.

- This addresses #92 
- and also adds `declarativeUpdates` for #94 that restarts VMs when the toplevel is changed and the host is `nixos-rebuild --switch`-ed

This PR also renames an option which might or might not be desirable. I'm not overly familiar with the codebase yet, so I "hacked" together something to make the microvm list command work again, since it depended on the `flake` file to determine the vm's toplevel. This should probably be revised, feedback is always welcome.

Example usage:

```nix
{ microvm, pkgs, ... }: {
  microvm.declarativeUpdates = true;
  microvm.restartIfChanged = true;
  microvm.vms.test.config = {
    imports = [microvm.nixosModules.microvm];
    # typical config ...
  };
}
```

This also allows merging multiple definitions, so specifying `config` again in another module will just work:

```nix
{
  microvm.vms.test.config = {
    services.nginx.enable = true;
  };
}
```

You can also specify another `pkgs` set and `specialArgs`, in case the guest should use a different system, or needs certain special arguments:


```nix
{ microvm, pkgs, ... }: {
  microvm.vms.test = {
    inherit pkgs; # optional (default to host's pkgs), used to selects the guest system and nixpkgs package set.
    specialArgs = {}; # optional, passed to nixosSystem
    config = {
      imports = [microvm.nixosModules.microvm];
      # typical config ...
    };
  };
}
```

Remarks:

All outstanding issues have been addressed. See comments below.

- ~~currently an assertions ensures `delcarativeUpdates -> restartIfChanged`. This could maybe be merged into one option, or left out by implying declarativeUpdates when using `vms.<name>.config`.~~
- ~~I'm not familiar enough with all the systemd services that are created, so I'm not 100% confident that `restartIfChanged` is always doing what it should be doing~~
- ~~declarative vm's are not yet restarted when their toplevel is changed. Waiting on #94 to see how we will solve that.~~

astro commented:
Cool, thanks. I guess this would make this project usable to even more contributors. :tada: 

Did you consider extending the non-flake story to imperative microvm management with the `microvm` command, and maybe even packages? No need to bloat this PR just yet, I going to review as is...

oddlama commented:
> Did you consider extending the non-flake story to imperative microvm management with the `microvm` command, and maybe even packages? No need to bloat this PR just yet, I going to review as is...

Sure, but I wasn't sure what the preferred implementation would be. Since I'm not overly familiar with the internals of microvm yet I refrained from implementing more than necessary for now. But I'm open to extending this. In either case I'd like to address the remarks before considering this final. Is there a matrix room for microvm yet?

oddlama commented:
I've now restructured this a bit to resolve some of the outstanding issues:

- The `declarativeUpdates` option was removed. Fully-declarative microvms are now distinguished from declaratively deployed but imperatively manged ones simply by checking whether they used `config` or `flake` in the definition.
- Moved the restartIfChanged option to a per-vm configuration. Purely imperative microvms couldn't be affected by this anyway so this is more granular and more correct.
- The `restartIfChanged` directive has been removed from many units, since `partOf` already accomplishes the same. Instead the master group unit `microvm@<name>` is now restarted if necessary.
- The microvm command will refuse to update fully-declarative vms.
- Added docs :D

This should address all your comments and my original remarks. Handling of the master unit restart is still accomplished by a hacky practice IMO, but I cannot think of another way to do it - suggestions are welcome.

oddlama commented:
I'd say this is ready now, but for the automatic restarts to take effect we will have to wait for https://github.com/NixOS/nixpkgs/pull/232230.

This PR adds deterministic restart on switch behavior for declarative MicroVMs, but not for imperative MicroVMs. If we want users to be able to select whether their imperative microvms should be restarted if the base service changes, this will have to be added as a separate per-instance template unit override. Probably the `microvm` command should create that drop-in, but this is out of scope for this PR.

astro commented:
I rebased this myself and added a few commits. Thank you!

oddlama commented:
Damn, now the GPG signatures are lost 😛 
But thanks for working on this ❤️
